ocaml/typing/typetexp.ml

```diff
@@ -786,41 +820,56 @@ and transl_type_aux env ~row_context ~aliased ~policy mode styp =
                 no statistically significant increase in build time. *)
              let reason = Jkind.History.Imported_type_argument
                             {parent_path = path; position = idx + 1; arity} in
-             Types.set_var_jkind ty' (Jkind.History.update_reason jkind reason)
-           | _ -> ()
-           end;
-           try unify_param env ty' cty.ctyp_type with Unify err ->
+             let jkind = Jkind.History.update_reason jkind reason in
+             Types.set_var_jkind ty' jkind;
+             jkind
+           | _ -> type_jkind_purely env ty'
+           in
+           let cty =
+            transl_type env ~policy ~jkind_check:(Exact jkind)
+              ~row_context Alloc.Const.legacy sty
+           in
+           (try unify_param env ty' cty.ctyp_type with Unify err ->
              let err = Errortrace.swap_unification_error err in
-             raise (Error(sty.ptyp_loc, env, Type_mismatch err))
+             raise (Error(sty.ptyp_loc, env, Type_mismatch err)));
+           cty
         )
-        (List.combine (List.combine stl args) params);
+        (List.combine stl params)
+      in
       let constr =
         newconstr path (List.map (fun ctyp -> ctyp.ctyp_type) args) in
       ctyp (Ttyp_constr (path, lid, args)) constr
   | Ptyp_app (st, stl) ->
-    let ty = transl_type env ~policy ~row_context Alloc.Const.legacy st in
-    let arg_tys =
-      List.map (transl_type env ~policy ~row_context Alloc.Const.legacy) stl
-    in
-    let app_jkind = estimate_type_jkind env ty.ctyp_type in
-    let app_err violation =
-      let arg_jkinds =
-        List.map (fun arg -> estimate_type_jkind env arg.ctyp_type) arg_tys
-      in
-      Error (
+    let ty =
+      transl_type env ~policy
+        ~jkind_check:(Arity (List.length stl, jkind_check))
+        ~row_context Alloc.Const.legacy st in
+    let ty_jkind = estimate_type_jkind env ty.ctyp_type in
```

Comment:
Using `estimate_type_jkind` here will result in the inference being order dependent. For instance, consider this example:
```ocaml
type ('f : any => value) any_to_value
val foo : 'f any_to_value -> 'a 'f -> unit
```
This will look at `'f any_to_value` before it looks at `'a 'f` which will result in `'a` have kind `any`. (I'm assuming, I think correctly, for these examples that `val` uses the `any` policy). But if you change it to:
```ocaml
type ('f : any => value) any_to_value
val foo : 'a 'f -> 'f any_to_value -> unit
```
then it will look at `'a 'f` first and default `'f` to have a sort variable for its argument, which will then become the kind of `'a`.

Instead of using `estimate_type_jkind` we should instead only use information that comes directly from the type expression being applied. That means we can use `estimate_type_jkind` for `Typ_constr` but for `Typ_var` we should only be using information coming from annotations on the binding of the variable.